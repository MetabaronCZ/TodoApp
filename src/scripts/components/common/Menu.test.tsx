import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { Menu, MenuItem } from 'components/common/Menu';

import { theme } from 'modules/theme';
import { hexToRgb } from 'modules/colors';
import { withMockedProviders } from 'test/component';

describe('components/common/Menu', () => {
  it('should render correctly', () => {
    const testData: MenuItem[] = [{ id: '1', title: 'CONTENT' }];

    const tree = render(withMockedProviders(<Menu items={testData} />));
    expect(tree.container).toMatchSnapshot();
    expect(tree.getByText('CONTENT')).toBeInTheDocument();
  });

  it('should render nothing when no children', () => {
    const { container } = render(withMockedProviders(<Menu items={[]} />));
    expect(container).toMatchSnapshot();
    expect(container.innerHTML).toEqual('');
  });

  it('should render menu items', () => {
    const testData: MenuItem[] = [
      { id: '1', title: 'Item 1', href: '/test' },
      { id: '2', title: 'Item 2', active: true },
    ];
    const tree = render(withMockedProviders(<Menu items={testData} />));
    expect(tree.container).toMatchSnapshot();

    expect(tree.getByRole('list')).toBeInTheDocument();
    expect(tree.getByText('Item 1')).toBeInTheDocument();
    expect(tree.getByText('Item 2')).toBeInTheDocument();

    expect(tree.getAllByRole('link').length).toEqual(1);
    expect(tree.getByRole('link')).toHaveAttribute('href', '/test');

    expect(tree.getAllByRole('button').length).toEqual(1);
    expect(tree.getByRole('button')).not.toHaveAttribute('href');

    const items = tree.container.querySelectorAll('a, button');

    expect(items[0]).not.toHaveStyle({
      background: hexToRgb(theme.color.active),
    });
    expect(items[1]).toHaveStyle({
      background: hexToRgb(theme.color.active),
    });
  });

  it('should call back onClick event', () => {
    const onClick = vi.fn();
    expect(onClick).not.toBeCalled();

    const testData: MenuItem[] = [{ id: '1', title: 'Item 1', onClick }];
    const tree = render(withMockedProviders(<Menu items={testData} />));

    fireEvent.click(tree.getByRole('button'));
    expect(onClick).toBeCalledTimes(1);
    expect(onClick).lastCalledWith();
  });
});
