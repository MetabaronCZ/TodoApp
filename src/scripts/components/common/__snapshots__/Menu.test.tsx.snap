// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/common/Menu > should render correctly 1`] = `
.c2 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  background: #f0f0f0;
}

.c3:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c3:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c3:disabled {
  color: #c0c0c0;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  border-bottom: 1px solid #c0c0c0;
}

.c1:last-child {
  border-bottom: none;
}

.c4 {
  width: 100%;
  text-align: left;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <button
        class="c2 c3 c4"
        type="button"
      >
        <span
          class="c5"
        >
          CONTENT
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`components/common/Menu > should render menu items 1`] = `
.c5 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c6 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  background: #f0f0f0;
}

.c6:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c6:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c6:disabled {
  color: #c0c0c0;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  border-bottom: 1px solid #c0c0c0;
}

.c1:last-child {
  border-bottom: none;
}

.c3 {
  width: 100%;
  text-align: left;
}

.c7 {
  width: 100%;
  background: #c0c0c0;
  text-align: left;
}

.c7:hover {
  background-color: #c0c0c0;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/test"
      >
        <span
          class="c4"
        >
          Item 1
        </span>
      </a>
    </li>
    <li
      class="c1"
    >
      <button
        class="c5 c6 c7"
        type="button"
      >
        <span
          class="c4"
        >
          Item 2
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`components/common/Menu > should render nothing when no children 1`] = `<div />`;
